src/Adapter/Cart/CommandHandler/UpdateCartCurrencyHandler.php

```diff
@@ -0,0 +1,117 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Cart\CommandHandler;
+
+use Cart;
+use Currency;
+use PrestaShop\PrestaShop\Core\Domain\Cart\Command\UpdateCartCurrencyCommand;
+use PrestaShop\PrestaShop\Core\Domain\Cart\CommandHandler\UpdateCartCurrencyHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Cart\Exception\CartException;
+use PrestaShop\PrestaShop\Core\Domain\Cart\ValueObject\CartId;
+use PrestaShop\PrestaShop\Core\Domain\Currency\Exception\CurrencyException;
+use PrestaShop\PrestaShop\Core\Domain\Currency\Exception\CurrencyNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\Currency\ValueObject\CurrencyId;
+
+/**
+ * @internal
+ */
+final class UpdateCartCurrencyHandler implements UpdateCartCurrencyHandlerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(UpdateCartCurrencyCommand $command)
+    {
+        $currency = $this->getCurrencyObject($command->getNewCurrencyId());
+
+        $this->assertCurrencyCantBeUsedInCart($currency);
+
+        $cart = $this->getCartObject($command->getCartId());
+        $cart->id_currency = (int) $currency->id;
+
+        if (false === $cart->save()) {
+            throw new CartException('Failed to update cart currency.');
+        }
+
+        // @todo: Should context be changed at controller layer instead?
+        \Context::getContext()->currency = $currency;
+        \Context::getContext()->cart = $cart;
```

Comment:
I think we should have some kind of `CartContext` service in controller layer, @matks wdyt it?

Reply:
If all the code that couples Cart and Context was in legacy controller (so we remove it) we can do it. But if there is some code doing that in core Cart class, we cannot move it else we're going to have half the behavior coupled and the other half decoupled 😒. Does it make sense ?

Reply:
I guess so.